### PR TITLE
Feat/neo-rpc: Retrive all methods in the module into a single list with getNodesByX

### PR DESCRIPTION
**Pull request by phetter**
- getNodesBy.all()
    - This can pull version, height, mempool, and connections into a single list sortable by node URL or any other value.

- Update notes for query.js to provide a reference link to API RPC method calls.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -100,6 +100,7 @@ See src/nodejs/ for the following:
   * Automatically select nodes
   * Get nodes by configurable sort factor
   * GetNodesByX
+    * Be careful, this can produce a lot of node traffic. It first pings each node in the list generated or provided to make sure they are up and within operating parameters and then calls the respective method requested. See [neo-rpc](#neo-rpc) for examples.
```
Comment (DalderupMaurice) [inline README.md]: Should there be a confirmation prompt before executing this call with this message? Many people won't read this

File: src/nodejs/neo-rpc/v2.9.0/client/cli/query.js [hunk 2]
```diff
@@ -38,11 +38,12 @@ program
   .option('-T, --Txs', 'Only return an array of transactions for the block')
   .option('-H, --Human', 'I am human so make outputs easy for human')
   .option('-N, --Net [Net]', 'Select network [net]: i.e., TestNet or MainNet', 'TestNet')
-  .option('-m, --method [method]', 'Call node with this RPC method, default \'getBlockCount\'', 'getblockcount')
+  .option('-m, --method [method]', 'Call node with this RPC method, default \'getblockcount\'', 'getblockcount')
   .option('-p, --params [params]', 'Call RPC method with these params, default is blank', '')
   .on('--help', function(){
-    print("Note: Currently, arguments that modify the results of an RPC call are NOT IMPLEMENTED.")
-    print("      It Is highly recommended to use neo-rpc/client/cli/getNodesByX to find a node to use and then use this programs --node or -n option")
+    print('Note: Currently, arguments that modify the results of an RPC call are NOT IMPLEMENTED.')
```
Comment (DalderupMaurice) [inline src/nodejs/neo-rpc/v2.9.0/client/cli/query.js]: You might want to add a TODO here

File: src/nodejs/neo-rpc/v2.9.0/client/module/getNodesBy.js [hunk 3]
```diff
@@ -342,3 +342,58 @@ exports.rawmempool = (options) => {
...
+    operations.forEach(op => {
+      op(options).then(rankedNodes => {
+        list[op.keyName] = rankedNodes
+
+        if (i++ === operations.length-1) {
+          operations.forEach(op => {
+            print('retrieving op: ' + op.keyName)
+
+            list[op.keyName].forEach(node => {
+              operations.forEach(op2 => {
```
Comment (DalderupMaurice) [inline src/nodejs/neo-rpc/v2.9.0/client/module/getNodesBy.js]: why are you looping 3 times over the same object `operations?`
This seems redundant as op, op and op2 will return the same results

File: src/nodejs/neo-rpc/v2.9.0/client/module/getNodesBy.js [hunk 4]
```diff
@@ -342,3 +342,58 @@ exports.rawmempool = (options) => {
...
+
+  let operations = [getTallest, getConnections, getVersion, getRawMemPool]
+
+  return new Promise((resolve, reject) => {
+    operations.forEach(op => {
+      op(options).then(rankedNodes => {
+        list[op.keyName] = rankedNodes
+
+        if (i++ === operations.length-1) {
+          operations.forEach(op => {
```
Comment (DalderupMaurice) [inline src/nodejs/neo-rpc/v2.9.0/client/module/getNodesBy.js]: `op` is already declared in an upper scope

File: src/nodejs/neo-rpc/v2.9.0/client/module/getNodesBy.js [hunk 5]
```diff
@@ -342,3 +342,58 @@ exports.rawmempool = (options) => {
...
+  getConnections.keyName = 'connections'
+  let getVersion = this.version
+  getVersion.keyName = 'version'
+  let getRawMemPool = this.rawmempool
+  getRawMemPool.keyName = 'rawmempool'
+
+  let operations = [getTallest, getConnections, getVersion, getRawMemPool]
+
+  return new Promise((resolve, reject) => {
+    operations.forEach(op => {
```
Comment (DalderupMaurice) [inline src/nodejs/neo-rpc/v2.9.0/client/module/getNodesBy.js]: a `forEach` has a 2nd parameter with the index. No need to keep track of `i` and `j`

File: README.md (hunk 1 above)
Comment (phetter) [reply README.md]: Excellent idea! Thanks

File: src/nodejs/neo-rpc/v2.9.0/client/cli/query.js (hunk 2 above)
Comment (phetter) [reply src/nodejs/neo-rpc/v2.9.0/client/cli/query.js]: There already is one at the top of the file.

File: src/nodejs/neo-rpc/v2.9.0/client/module/getNodesBy.js (hunk 4 above)
Comment (phetter) [reply src/nodejs/neo-rpc/v2.9.0/client/module/getNodesBy.js]: The outermost loop collects the data. The innermost compares it.

File: src/nodejs/neo-rpc/v2.9.0/client/module/getNodesBy.js (hunk 3 above)
Comment (phetter) [reply src/nodejs/neo-rpc/v2.9.0/client/module/getNodesBy.js]: The outermost loop collects the data. The innermost compares it.

**Comment by phetter** [comment]
Thanks for the great review, as always @DalderupMaurice!

**Comment by jeroenptrs** [comment]
@phetter It feels like I'm missing something, there's some massive performance optimizations possible that @DalderupMaurice hinted at. Honestly, I suggest you review the looping you do in getNodesBy because it's far from optimal.

Take a look in the Array.prototype specifications on MDN https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/prototype

**Pull request merged**